### PR TITLE
Document IREE-specific CMake options

### PR DESCRIPTION
**Pull request by marbre**
First attempt to solve #2111. There is still room to improve the layout.

### CODE REVIEW

File: docs/GetStarted/cmake_options_and_variables.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
+# IREE-specific CMake Options and Variables
+
+This gives a brief explanation of IREE specific CMake options and variables.
```
Comment (ScottTodd) [inline docs/GetStarted/cmake_options_and_variables.md:3]: We could expand on this overview section, taking inspiration from https://llvm.org/docs/CMake.html
(that doesn't have to happen in this PR)

File: docs/GetStarted/cmake_options_and_variables.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+* `IREE_ENABLE_RUNTIME_TRACING`:BOOL
+
+  Enables instrumented runtime tracing. Defaults to `OFF`.
```
Comment (ScottTodd) [inline docs/GetStarted/cmake_options_and_variables.md:7]: I'd like to iterate on the layout/styling, but that doesn't have to happen in this PR.

Some ideas:

* Use headers (``### `IREE_ENABLE_RUNTIME_TRACING`:BOOL``) so we can link to individual options
* Drop bullet points
* Adjust indentation (see https://llvm.org/docs/CMake.html)
* Try using a table and some CSS
* Organize into sections (which components to build, which backends to enable, how to configure dependencies, etc.)

File: docs/GetStarted/cmake_options_and_variables.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+
+* `IREE_BUILD_EXPERIMENTAL`:BOOL
+
+  Builds experimental projects. Defaults to `OFF`.
+
+
+* `IREE_HAL_DRIVERS_TO_BUILD`:STRING
+
+  *This does not has any effect at the moment, but will be supported in the future!*
```
Comment (ScottTodd) [inline docs/GetStarted/cmake_options_and_variables.md:56]:
```suggestion
  *This does not have any effect at the moment, but will be supported in the future!*
```
nit: has -> have

File: docs/GetStarted/cmake_options_and_variables.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  Semicolon-separated list of HAL drivers to build, or `all` for building all HAL drivers. Case-insensitive. Defaults to `all`. Example: `-DIREE_HAL_DRIVERS_TO_BUILD="Vulkan;VMLA"`.
+
+
+* `IREE_TARGET_BACKENDS_TO_BUILD`:STRING
+
+  *This does not has any effect at the moment, but will be supported in the future!*
```
Comment (ScottTodd) [inline docs/GetStarted/cmake_options_and_variables.md:62]:
```suggestion
  *This does not have any effect at the moment, but will be supported in the future!*
```
nit: has -> have

File: docs/GetStarted/cmake_options_and_variables.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+
+* `IREE_ENABLE_LLD`:BOOL
+
+  Use lld when linking. Defaults to `OFF`.
+
+
+* `IREE_MLIR_DEP_MODE`:STRING
+
+  Defines the MLIR dependency mode. Case-sensitive. Can be `BUNDELED`, `DISABLED` or `INSTALLED`. Defaults to `INSTALLED`.
```
Comment (ScottTodd) [inline docs/GetStarted/cmake_options_and_variables.md:73]:
```suggestion
  Defines the MLIR dependency mode. Case-sensitive. Can be `BUNDLED`, `DISABLED` or `INSTALLED`. Defaults to `INSTALLED`.
```
typo

**Comment by antiagainst** [comment]
This is nice, thanks @marbre!

A general comment: it would be nice to flesh out the explanation for each option a bit more. At the moment I feel it's basically the message embedded in `CMakeLists.txt`. 

Also I think we need to copy the newly added file over in https://github.com/google/iree/blob/master/build_tools/cmake/build_docs.sh.

File: docs/GetStarted/cmake_options_and_variables.md (hunk 5 above)
Comment (marbre) [reply docs/GetStarted/cmake_options_and_variables.md]: Fixed and further corrected the default which is `BUNDLED` and not `INSTALLED`.

**Comment by marbre** [comment]
> This is nice, thanks @marbre!
> 
> A general comment: it would be nice to flesh out the explanation for each option a bit more. At the moment I feel it's basically the message embedded in `CMakeLists.txt`.

I fully agree! So far the explanations for most options do not go beyond what already is in the `CMakeLists.txt`. The main motivation so far was to add some words on `IREE_MLIR_DEP_MODE`, which was lacking the info that LLVM needs to be compiled with RTTI enabled. Instead of only adding this information somewhere, I started to document all options, but we definitely need to improve this. So this is more a starting point and does not resolve #2111.

> Also I think we need to copy the newly added file over in https://github.com/google/iree/blob/master/build_tools/cmake/build_docs.sh.

Missed that. Thanks for pointing to this :)

File: docs/GetStarted/cmake_options_and_variables.md (hunk 2 above)
Comment (marbre) [reply docs/GetStarted/cmake_options_and_variables.md:7]: Yes, I fully agree. I don't really like the current layout myself and we need to iterate on it. This is only a first quick start to document the options.

File: docs/GetStarted/cmake_options_and_variables.md (hunk 1 above)
Comment (marbre) [reply docs/GetStarted/cmake_options_and_variables.md:3]: Yes, I was looking into that one before starting (also I don't like the layout :wink:), but I wasn't sure how far we should document the options. So far we only document the IREE specific ones, somehow equivalent to https://llvm.org/docs/CMake.html#llvm-specific-variables. In addition we have one MLIR specific variable which is `MLIR_DIR`. For that one introducing a new section seemed overloaded. But indeed, we can create sections like https://llvm.org/docs/CMake.html#frequently-used-cmake-variables and could further add explanations or links to the appropriate docs for options like `CMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER`.

File: docs/GetStarted/cmake_options_and_variables.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  Builds the IREE compiler. Defaults to `ON`.
+
+
+* `IREE_BUILD_TESTS`:BOOL
+
+  Builds IREE unit tests. Defaults to `ON`.
+
+* `IREE_BUILD_DOCS`:BOOL
+
+  Builds IREE docs. Defaults to `OFF`.
```
Comment (ScottTodd) [inline docs/GetStarted/cmake_options_and_variables.md:31]:
```suggestion
  Builds IREE documentation. Defaults to `OFF`.
```
test suggested change

**Comment by ScottTodd** [comment]
```
Conflicting files
scripts/prepare_doc_publication.py
```
Resolve and merge? Or do you have more comments, @antiagainst ?

**Comment by marbre** [comment]
> ```
> Conflicting files
> scripts/prepare_doc_publication.py
> ```
> 
> Resolve and merge? Or do you have more comments, @antiagainst ?

Conflict resolved :)

**Pull request closed without merging**